llpc/tool/llpcAutoLayout.cpp

```diff
@@ -77,6 +80,211 @@ static unsigned getTypeDataSize(const SPIRVType *ty) {
+
+    if (userDataNode->node.type == Llpc::ResourceMappingNodeType::DescriptorTableVaPtr) {
+      if (userDataNode->node.tablePtr.pNext[0].srdRange.set == set) {
+        descriptorTableVaPtr = userDataNode;
+        break;
+      }
+    }
+  }
+
+  return descriptorTableVaPtr;
+}
+
+// =====================================================================================================================
+// Find userDataNode with specified set and binding. And return Node index.
+//
+// @param userDataNode : ResourceMappingNode pointer
+// @param nodeCount : User data node count
+// @param set : Find same set in node array
+// @param binding : Find same binding in node array
+// @param [out] index : Return node position in node array
+static const ResourceMappingNode *findResourceNode(const ResourceMappingNode *userDataNode, unsigned nodeCount,
+                                                   unsigned set, unsigned binding, unsigned *index) {
+  const ResourceMappingNode *resourceNode = nullptr;
+
+  for (unsigned j = 0; j < nodeCount; ++j) {
+    const ResourceMappingNode *next = &userDataNode[j];
+
+    if (set == next->srdRange.set && binding == next->srdRange.binding) {
+      resourceNode = next;
+      *index = j;
+      break;
+    }
+  }
+
+  return resourceNode;
+}
+
+// =====================================================================================================================
+// Find userDataNode with specified set and binding. And return Node index.
+//
+// @param userDataNode : ResourceMappingRootNode pointer
+// @param nodeCount : User data node count
+// @param set : Find same set in node array
+// @param binding : Find same binding in node array
+// @param [out] index : Return node position in node array
+// @returns : The Node index
+static const ResourceMappingRootNode *findResourceNode(const ResourceMappingRootNode *userDataNode, unsigned nodeCount,
+                                                       unsigned set, unsigned binding, unsigned *index) {
+  const ResourceMappingRootNode *resourceNode = nullptr;
+
+  for (unsigned j = 0; j < nodeCount; ++j) {
+    const ResourceMappingRootNode *next = &userDataNode[j];
+
+    if (set == next->node.srdRange.set && binding == next->node.srdRange.binding) {
+      resourceNode = next;
+      *index = j;
+      break;
+    }
+  }
+
+  return resourceNode;
+}
+
+// =====================================================================================================================
+// Check if autoLayoutUserDataNodes is a subset of userDataNodes.
+//
+// @param [in] resourceMapping : Resource mapping data, which can contain user data nodes
+// @param [in] autoLayoutUserDataNodeCount : UserData Node count
+// @param [in] autoLayoutUserDataNodes : ResourceMappingNode
+// @returns : true if compatible
+bool checkResourceMappingComptible(const ResourceMappingData *resourceMapping, unsigned autoLayoutUserDataNodeCount,
```

Comment:
This function is very long. It should be split into smaller helper functions to make it more maintainable. It also deserved unit tests.